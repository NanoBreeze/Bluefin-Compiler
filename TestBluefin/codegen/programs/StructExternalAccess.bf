/*
Verifies external memberAccess of struct variables.
TODO: verify method calls
TODO: verify return values of method calls too. eg) f().g().h(); or f().a;
TODO: refactor the implementation. Right now, primaryIds whose resolved symbol has type of StructSymbol
and intermediate member access both create "load" instructions, which isn't necessary
*/

struct Foo {
	int x;
	float y;
};

struct Bar {
	Foo foo;
	int a;
	float b;
};

void foo() {
	Foo foo;
	foo.y;				// single member access
	foo.x + 5;			// using member access in expr
	foo.y = foo.x - 5;	// simple assignment with member access
}


void bar() {
	Bar bar;
	bar.foo;			// member access where member is a variable with struct type

	Bar barTwo;
	barTwo.foo;			// check that we're referring to barTwo's addr and not bar's

	bar.foo.x;			// chained member access

	bar.foo.y = 10; //bar.foo.x - bar.foo.y;		// chained member access used in expr and simple assignment
	bar.foo.y = bar.foo.x - bar.foo.y;		// chained member access used in expr and simple assignment
}

