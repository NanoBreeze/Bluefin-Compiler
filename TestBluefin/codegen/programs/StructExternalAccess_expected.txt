; ModuleID = 'StructExternalAccess.bf'
source_filename = "StructExternalAccess.bf"

%Foo = type { i32, float }
%Bar = type { %Foo, i32, float }

define void @_ctor_Foo(%Foo* %this) {
entry:
  ret void
}

define void @_ctor_Bar(%Bar* %this) {
entry:
  %memPtr_foo = getelementptr %Bar, %Bar* %this, i32 0, i32 0
  call void @_ctor_Foo(%Foo* %memPtr_foo)
  ret void
}

define void @foo() {
entry:
  %alloctmp_foo = alloca %Foo, align 8
  call void @_ctor_Foo(%Foo* %alloctmp_foo)
  %loadtmp_foo = load %Foo, %Foo* %alloctmp_foo, align 4
  %memPtr_y = getelementptr %Foo, %Foo* %alloctmp_foo, i32 0, i32 1
  %loadtmp_y = load float, float* %memPtr_y, align 4
  %loadtmp_foo1 = load %Foo, %Foo* %alloctmp_foo, align 4
  %memPtr_x = getelementptr %Foo, %Foo* %alloctmp_foo, i32 0, i32 0
  %loadtmp_x = load i32, i32* %memPtr_x, align 4
  %addtmp = add i32 %loadtmp_x, 5
  %loadtmp_foo2 = load %Foo, %Foo* %alloctmp_foo, align 4
  %memPtr_y3 = getelementptr %Foo, %Foo* %alloctmp_foo, i32 0, i32 1
  %loadtmp_y4 = load float, float* %memPtr_y3, align 4
  %loadtmp_foo5 = load %Foo, %Foo* %alloctmp_foo, align 4
  %memPtr_x6 = getelementptr %Foo, %Foo* %alloctmp_foo, i32 0, i32 0
  %loadtmp_x7 = load i32, i32* %memPtr_x6, align 4
  %subtmp = sub i32 %loadtmp_x7, 5
  %casttmp = sitofp i32 %subtmp to float
  store float %casttmp, float* %memPtr_y3, align 4
  ret void
}

define void @bar() {
entry:
  %alloctmp_bar = alloca %Bar, align 8
  call void @_ctor_Bar(%Bar* %alloctmp_bar)
  %loadtmp_bar = load %Bar, %Bar* %alloctmp_bar, align 4
  %memPtr_foo = getelementptr %Bar, %Bar* %alloctmp_bar, i32 0, i32 0
  %loadtmp_foo = load %Foo, %Foo* %memPtr_foo, align 4
  %alloctmp_barTwo = alloca %Bar, align 8
  call void @_ctor_Bar(%Bar* %alloctmp_barTwo)
  %loadtmp_barTwo = load %Bar, %Bar* %alloctmp_barTwo, align 4
  %memPtr_foo1 = getelementptr %Bar, %Bar* %alloctmp_barTwo, i32 0, i32 0
  %loadtmp_foo2 = load %Foo, %Foo* %memPtr_foo1, align 4
  %loadtmp_bar3 = load %Bar, %Bar* %alloctmp_bar, align 4
  %memPtr_foo4 = getelementptr %Bar, %Bar* %alloctmp_bar, i32 0, i32 0
  %loadtmp_foo5 = load %Foo, %Foo* %memPtr_foo4, align 4
  %memPtr_x = getelementptr %Foo, %Foo* %memPtr_foo4, i32 0, i32 0
  %loadtmp_x = load i32, i32* %memPtr_x, align 4
  %loadtmp_bar6 = load %Bar, %Bar* %alloctmp_bar, align 4
  %memPtr_foo7 = getelementptr %Bar, %Bar* %alloctmp_bar, i32 0, i32 0
  %loadtmp_foo8 = load %Foo, %Foo* %memPtr_foo7, align 4
  %memPtr_y = getelementptr %Foo, %Foo* %memPtr_foo7, i32 0, i32 1
  %loadtmp_y = load float, float* %memPtr_y, align 4
  store float 1.000000e+01, float* %memPtr_y, align 4
  %loadtmp_bar9 = load %Bar, %Bar* %alloctmp_bar, align 4
  %memPtr_foo10 = getelementptr %Bar, %Bar* %alloctmp_bar, i32 0, i32 0
  %loadtmp_foo11 = load %Foo, %Foo* %memPtr_foo10, align 4
  %memPtr_y12 = getelementptr %Foo, %Foo* %memPtr_foo10, i32 0, i32 1
  %loadtmp_y13 = load float, float* %memPtr_y12, align 4
  %loadtmp_bar14 = load %Bar, %Bar* %alloctmp_bar, align 4
  %memPtr_foo15 = getelementptr %Bar, %Bar* %alloctmp_bar, i32 0, i32 0
  %loadtmp_foo16 = load %Foo, %Foo* %memPtr_foo15, align 4
  %memPtr_x17 = getelementptr %Foo, %Foo* %memPtr_foo15, i32 0, i32 0
  %loadtmp_x18 = load i32, i32* %memPtr_x17, align 4
  %casttmp = sitofp i32 %loadtmp_x18 to float
  %loadtmp_bar19 = load %Bar, %Bar* %alloctmp_bar, align 4
  %memPtr_foo20 = getelementptr %Bar, %Bar* %alloctmp_bar, i32 0, i32 0
  %loadtmp_foo21 = load %Foo, %Foo* %memPtr_foo20, align 4
  %memPtr_y22 = getelementptr %Foo, %Foo* %memPtr_foo20, i32 0, i32 1
  %loadtmp_y23 = load float, float* %memPtr_y22, align 4
  %subtmp = fsub float %casttmp, %loadtmp_y23
  store float %subtmp, float* %memPtr_y12, align 4
  ret void
}
